File: solutions/fortran/protein-translation/protein_translation.f90
module protein_translation
   implicit none
   character(len=4) :: StopCodon = "STOP"
   private
   public :: proteins
contains

   pure function proteins(rna) result(names)
      character(len=*), intent(in) :: rna
      character(len=13), allocatable :: names(:), temp(:)
      character(len=3) :: codon
      integer :: i, n, valid

      valid = 0
      n = len(rna) / 3
      allocate(temp(n))

      do i = 1,n
         codon = rna(i*3-2:i*3)
         if ( from(codon) == StopCodon ) then
            exit
         end if
         valid = valid + 1
         temp(i) = from(codon)
      end do

      if ( valid > 0 ) then
         allocate(names(valid))
         names = temp(1:valid)
      else
         names = [character(len=13) :: ]
      end if
   end function proteins

   pure function from(codon) result(name)
      character(len=*), intent(in) :: codon
      character(len=13) :: name

      select case (codon)
       case ("AUG")
         name = "Methionine"
       case ("UUU", "UUC")
         name = "Phenylalanine"
       case ("UUA","UUG")
         name = "Leucine"
       case ("UCU", "UCC", "UCA", "UCG")
         name = "Serine"
       case ("UAU", "UAC")
         name = "Tyrosine"
       case ("UGU", "UGC")
         name = "Cysteine"
       case ("UGG")
         name = "Tryptophan"
       case ("UAA", "UAG", "UGA")
         name = StopCodon
      end select
   end function from
end module protein_translation

File: solutions/lua/protein-translation/protein-translation.lua
local c_stop_codon = "STOP"
local c_mapping = {
    UGU = "Cysteine",
    UGC = "Cysteine",
    UUA = "Leucine",
    UUG = "Leucine",
    AUG = "Methionine",
    UUU = "Phenylalanine",
    UUC = "Phenylalanine",
    UCU = "Serine",
    UCC = "Serine",
    UCA = "Serine",
    UCG = "Serine",
    UGG = "Tryptophan",
    UAU = "Tyrosine",
    UAC = "Tyrosine",
    UAA = c_stop_codon,
    UAG = c_stop_codon,
    UGA = c_stop_codon
}

local function translate_codon(codon)
    return c_mapping[codon] or error("invalid codon")
end

local function translate_rna_strand(rna_strand)
    local result = {}
    for chunk in rna_strand:gmatch("...") do
        local amino_acid = translate_codon(chunk)
        if amino_acid == c_stop_codon then
            break
        end
        table.insert(result, amino_acid)
    end
    return result
end

return {codon = translate_codon, rna_strand = translate_rna_strand}
